numpy/core/src/multiarray/ctors.c

```diff
@@ -55,7 +55,19 @@ fromstr_next_element(char **s, void *dptr, PyArray_Descr *dtype,
      * if an error occurs s is not changed
      */
     if (*s == e) {
-        /* Nothing read */
+        /* Nothing read, could be end of string or an error */
+        if (end != NULL) {
+            /* check if we read until the intended end */
+            if (*s <= end) {
+                return -2;
```

Comment:
Can you add a comment to the `next_element` typedef above explaining the return value semantics?

Reply:
Also, note you have one other option that may or may not be a good idea - you could `PyErr_SetNone(PyExc_StopIteration)` for EOF

Reply:
Using the below helper,these 13 lines become
```C
/* non-empty string left means we had an error */
return string_is_empty(string, end) ? -1 : -2;
```